components/job-orchestration/job_orchestration/scheduler/search/search_scheduler.py

```diff
@@ -210,32 +242,64 @@ def check_job_status_and_update_db(db_conn):
             continue
 
         if returned_results is not None:
-            new_job_status = SearchJobStatus.SUCCEEDED
+            new_job_status = SearchJobStatus.RUNNING
             for task_result_obj in returned_results:
                 task_result = SearchTaskResult.parse_obj(task_result_obj)
                 if not task_result.success:
                     task_id = task_result.task_id
                     new_job_status = SearchJobStatus.FAILED
                     logger.debug(f"Task {task_id} failed - result {task_result}.")
 
-            del active_jobs[job_id]
-
-            if set_job_status(db_conn, job_id, new_job_status, SearchJobStatus.RUNNING):
-                if new_job_status != SearchJobStatus.FAILED:
-                    logger.info(f"Completed job {job_id}.")
-                else:
-                    logger.info(f"Completed job {job_id} with failing tasks.")
+            if new_job_status != SearchJobStatus.FAILED:
+                max_num_results = job.search_config.max_num_results
+                # check if we have searched all archives
+                if len(job.remaining_archives_for_search) == 0:
+                    new_job_status = SearchJobStatus.SUCCEEDED
+                # check if we have reached max results
+                elif max_num_results > 0:
+                    results_cache_client = pymongo.MongoClient(results_cache_uri)
+                    results_cache_collection = results_cache_client.get_default_database()[job_id]
+                    results_count = results_cache_collection.count_documents({})
+                    if results_count >= max_num_results:
+                        result = list(
+                            results_cache_collection.find()
+                            .sort("timestamp", -1)
+                            .limit(max_num_results)
+                        )
+                        if result is not None:
+                            min_timestamp_in_top_results = result[-1]["timestamp"]
+                        else:
+                            min_timestamp_in_top_results = 0
+                        results_cache_client.close()
```

Comment:
Can we extract this into a method to simplify the code a bit?

Reply:
We can extract this into a method, but it does a lot of things (create a mongodb client, get document count, run a query) and require two arguments `max_num_results` and `max_timestamp_in_archive`. I'm not sure how to name this function.

Reply:
Oh, I just meant extracting the part which does the MongoDB query. So how about `get_min_timestamp_in_top_results`?